Pettingzoo wrapper changed to gymnasium interface

alexander-zap opened the pull request:
### Proposed change(s)

I adjusted the Unity Pettingzoo Wrapper to support the gymnasium interface (gymnasium>=0.25).

This includes:

- returning `terminated, truncated` on step
- returning `info` on reset
- accept `seed` and `options` on reset
- not accept `mode` on render

Additionall, I fixed interface problems with pettingzoo implementation:

- all agents return an observations synchronously (instead of having steps where only one agent returns an observation)
- in case the agent sends a TerminationStep *and* a DecisionStep after, the reward from the termination step should be taken (but for observations it's the other way around because of SB3 convention)
- `agents` attribute should store agents even if they have *just* finished (they should be removed the step after)
- fixed a bug where a continuous action vector was assigned len(action) * action[0]
- fixed double removal of an agent from _live_agents for ParallelEnv

Also I bumped the dependencies for pettingzoo to 1.22.0 (because this is there Gymnasium support started on their side).

### Useful links (Github issues, JIRA tickets, ML-Agents forum threads etc.)

This also could be helpful for: https://github.com/Unity-Technologies/ml-agents/issues/5890


### Types of change(s)

- [x] Bug fix
- [x] New feature
- [x] Code refactor
- [x] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the changelog (if applicable)
- [x] Updated the documentation (if applicable)
- [ ] Updated the migration guide (if applicable)

### Other comments

Commit:
Adjusted .pre-commit-config.yaml for Windows usage

Commit:
Fixed mypy issue

Commit:
Switched go gymnasium interface

- added dependency to gymnasium instead of gym
- adjusted Unity gym interface to return `truncated` on step
- adjusted Unity gym interface to accept `seed` and `options` on reset (not used)
- adjusted Unity gym interface to not accept `mode` on reset (not used anyway)
- relaxed dependencies to numpy, pettingzoo and Python version

Commit:
Created pyproject.toml to support installation of package via poetry (without publishing)

alexander-zap committed:
Merge pull request #1 from alexander-zap/local_poetry_support

Local poetry support

Commit:
Updated imports to use gymnasium instead of gym

Commit:
Merge branch 'gymnasium_interface' of https://github.com/alexander-zap/ml-agents into develop

Commit:
Adjusted pettingzoo environments to support gymnasium interface (reset returning obs and info; step returning terminated and truncated instead of done)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Unity-Technologies/ml-agents?pullRequest=6211) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/Unity-Technologies/ml-agents?pullRequest=6211) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: alexander-zap<br/>:x: a_zap<br/><hr/>**a_zap** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Unity-Technologies/ml-agents?pullRequest=6211) it.</sub>

Commit:
Adjusted pettingzoo environments to support gymnasium interface (reset returning obs and info; step returning terminated and truncated instead of done)

Commit:
Set continuous action space as dtype float64 instead of int32

Commit:
Fixed interface problems with pettingzoo implementation

- all agents return an observations synchronously (instead of having steps where only one agent returns an observation)
- in case the agent sends a TerminationStep *and* a DecisionStep after, the reward from the termination step should be taken (but for observations it's the other way around because of SB3 convention)
- `agents` attribute should store agents even if they have *just* finished (they should be removed the step after)
- fixed a bug where a continuous action vector was assigned len(action) * action[0]
- fixed double removal of an agent from _live_agents for ParallelEnv

alexander-zap commented:
Should also solve #6214

montplaisir commented:
Thank you for your submission!
Your changes were covered by PR #6309 - Migrate gym to gymnasium.
You are credited as co-author.
